src/d3a/d3a_core/sim_results/file_export_endpoints.py

```diff
@@ -344,76 +342,83 @@ def _specific_row(self, slot, market):
+
+            if child.children:
+                self._accumulate_devices(child)
+
+    def _accumulate_self_consumption(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.consumer_list:
+            self.total_self_consumption_wh += trade.offer.energy * 1000
+
+    def _accumulate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.producer_list and trade.buyer_origin in self.ess_list:
+            self.self_consumption_buffer_wh += trade.offer.energy * 1000
+
+    def _dissipate_self_consumption_buffer(self, trade):
+        if trade.seller_origin in self.ess_list:
+            if trade.buyer_origin in self.consumer_list and self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                    self.total_self_consumption_wh += trade.offer.energy * 1000
                 else:
-                    if trade.buyer is child.name:
-                        self._total_energy += trade.offer.energy
-
-    def _export_house_pv_self_consumption(self, area):
-        house_pv_device = list()
-        house_load_device = list()
-        trade_by_pv = 0
-        total_energy_bought = 0
-        traded_from_pv = 0
-        if not _is_house_node(area):
-            return
-        for child in area.children:
-            if isinstance(child.strategy, PVStrategy):
-                house_pv_device.append(child.name)
-            elif _is_load_node(child) or _is_prosumer_node(child):
-                house_load_device.append(child.name)
-
-        for markets in area.past_markets:
-            for trade in markets.trades:
-                # Total Electricity traded by house device
-                if trade.buyer in house_load_device:
-                    total_energy_bought += trade.offer.energy
-                # Electricity produced by PV
-                if trade.offer.seller in house_pv_device:
-                    trade_by_pv += trade.offer.energy
-                # PV electricity self_consumption
-                if trade.offer.seller in house_pv_device and trade.buyer in house_load_device:
-                    traded_from_pv += trade.offer.energy
-
-        if trade_by_pv != 0:
-            self_consumption_within_house = traded_from_pv / trade_by_pv
-        else:
-            self_consumption_within_house = 0
-        if total_energy_bought != 0:
-            self_sufficiency = traded_from_pv / total_energy_bought
-        else:
-            self_sufficiency = 0
-
-        return {"self_consumption_within_house": self_consumption_within_house,
-                "self_sufficiency": self_sufficiency}
+                    self.total_self_consumption_wh += self.self_consumption_buffer_wh
+                    self.self_consumption_buffer_wh = 0
+            elif (trade.buyer_origin not in self.consumer_list or
```

Comment:
This elif statement isnt captioning a different case than the if branch. 
What I mean is: if `trade.buyer_origin in self.consumer_list and self.self_consumption_buffer_wh > 0` is True, both branched can be entered. I think the first one will be taken, but i think you can simplify this function with less if then else statements.

Reply:
Umm, I don't think so, first `if` checks the `trade.buyer_origin in self.consumer_list` and the later checks if `trade.buyer_origin not in self.consumer_list`

Reply:
Ah yes! Thanks! This was a test ;-) . Haha, joking!

Reply:
Buut, `(trade.buyer_origin not in self.consumer_list or trade.buyer_origin not in self.ess_list)` is True for all devices right?

Reply:
And thats why this statement is tested later not at the first if clause

Reply:
In this block, I assume that you want to check if an external device buys energy from the storage, correct ? If this is the case, I think that 'or' statement is not appropriate, 'and' should be the correct one. I have probably misunderstood something here though.

Reply:
`OR` statement is inside the brackets that is testing if the internal ESS` energy buyer is not either of internal consumer OR internal storage?

We can discuss offline too for better understanding!

Reply:
Resolved after offline discussion.